cuda_next/include/detail/hierarchy_dimensions.hpp

```diff
@@ -0,0 +1,496 @@
+};
+
+template <typename BottomUnit, typename... Levels>
+struct hierarchy_dimensions_fragment;
+
+// If lowest unit in the hierarchy is thread, it can be considered a full hierarchy and not only a fragment
+template <typename... Levels>
+using hierarchy_dimensions = hierarchy_dimensions_fragment<thread_level, Levels...>;
+
+namespace detail
+{
+// Function to sometimes convince the compiler something is a constexpr and not really accessing runtime storage
+// Mostly a work around for what was addressed in P2280 (c++23) by leveraging a default constructor of extents
+template <typename T, size_t... Extents>
+constexpr _CCCL_HOST_DEVICE auto fool_compiler(const dimensions<T, Extents...>& ex)
+{
+  if constexpr (ex.rank_dynamic() == 0)
+  {
+    return dimensions<T, Extents...>();
+  }
+  else
+  {
+    return ex;
+  }
+}
+
+template <typename QueryLevel, typename Hierarchy>
+struct has_level_helper;
+
+template <typename QueryLevel, typename Unit, typename... Levels>
+struct has_level_helper<QueryLevel, hierarchy_dimensions_fragment<Unit, Levels...>>
+    : public std::disjunction<std::is_same<QueryLevel, typename Levels::level_type>...>
+{};
+
+// Is this needed?
+template <typename QueryLevel, typename... Levels>
+struct has_level_helper<QueryLevel, hierarchy_dimensions<Levels...>>
+    : public std::disjunction<std::is_same<QueryLevel, typename Levels::level_type>...>
+{};
+
+template <typename QueryLevel, typename Hierarchy>
+struct has_unit
+{};
+
+template <typename QueryLevel, typename Unit, typename... Levels>
+struct has_unit<QueryLevel, hierarchy_dimensions_fragment<Unit, Levels...>> : std::is_same<QueryLevel, Unit>
+{};
+
+template <unsigned int Id, typename... Levels>
+using level_at_index = typename ::cuda::std::tuple_element<Id, ::cuda::std::tuple<Levels...>>::type;
+
+template <typename QueryLevel>
+struct get_level_helper
+{
+  template <typename TopLevel, typename... Levels>
+  constexpr auto& _CCCL_HOST_DEVICE operator()(const TopLevel& top, const Levels&... levels)
+  {
+    if constexpr (std::is_same_v<QueryLevel, typename TopLevel::level_type>)
+    {
+      return top;
+    }
+    else
+    {
+      return (*this)(levels...);
+    }
+  }
+};
+} // namespace detail
+
+template <typename QueryLevel, typename Hierarchy>
+bool constexpr has_level =
+  detail::has_level_helper<QueryLevel, std::remove_cv_t<std::remove_reference_t<Hierarchy>>>::value;
+
+template <typename QueryLevel, typename Hierarchy>
+bool constexpr has_level_or_unit =
+  detail::has_level_helper<QueryLevel, std::remove_cv_t<std::remove_reference_t<Hierarchy>>>::value
+  || detail::has_unit<QueryLevel, std::remove_cv_t<std::remove_reference_t<Hierarchy>>>::value;
+
+// We can consider removing the operator, but its convinient for in-line construction
+// TODO accept forwarding references
+template <typename LUnit, typename L1, typename... Levels>
+auto constexpr operator&(const hierarchy_dimensions_fragment<LUnit, Levels...>& ls, const L1& l1) noexcept
```

Comment:
This is evil, overloading `operator&` is almost always not what we want. I would need to see a good usecase for this that would not be served by one of the less dangerous operators

Reply:
I used & because it seemed ok semantically. Would operator* work better? I am not attached to any operator so if you have some suggestion I would be glad to hear it. I'm not even sure if we want the operator at all, but wanted to add it as optional and easier in-line hierarchy creation